registry/config.go

```diff
@@ -23,12 +23,46 @@ type ServiceOptions struct {
 	// V2Only controls access to legacy registries.  If it is set to true via the
 	// command line flag the daemon will not attempt to contact v1 legacy registries
 	V2Only bool `json:"disable-legacy-registry,omitempty"`
+
+	// Registries holds information associated with the specified registries.
+	Registries []Registry `json:"registries,omitempty"`
 }
 
 // serviceConfig holds daemon configuration for the registry service.
 type serviceConfig struct {
 	registrytypes.ServiceConfig
-	V2Only bool
+	V2Only     bool
+	Registries map[string]Registry
+}
+
+// Registry holds information for a registry and its mirrors.
+type Registry struct {
+	// Prefix is used for the lookup of endpoints, where the given registry
+	// is selected when its Prefix is a prefix of the passed reference, for
+	// instance, Prefix:"docker.io/opensuse" will match a `docker pull
+	// opensuse:tumleweed`.
+	Prefix string `json:"Prefix,omitempty"`
+	// The mirrors will be selected prior to the registry during lookup of
+	// endpoints.
+	Mirrors []Mirror `json:"Mirrors,omitempty"`
+	// Host is the host of the registry (e.g., "docker.io")
+	Host string
+	// This avoids to parse the URL of each new endpoint lookup.
+	URL *url.URL
+	// True if the host is "docker.io".
+	Official bool
+	// True if URL.Scheme == "https".
+	Secure bool
+}
+
+// Mirror holds information for a given registry mirror.
+type Mirror struct {
+	// Host is the URL specified by the user.  At deamon start, the passed
+	// URL will be loaded into Mirror.URL and corresponding host stored in
+	// Mirror.Host.
+	Host string `json:"URL"`
```

Comment:
Why is the JSON different from the field name?

Let's see if we can be consistent here.

Reply:
It's quite tricky to specify a valid `ul.URL` from JSON. Also, it wouldn't match the discussed prefix syntax. Maybe, the intention became more explicit by renaming `Host` to `Input` and add an extra `Host` member (if desired).

Reply:
The design problem is apparent in the doc message: "Host is the URL specified by the user".

This should be called URL or something like that, it if takes a URL. Extracting a host from a URL is cheap.

Reply:
I will address the issue in the next update for this PR. Thanks for your feedback!